packages/expo-modules-core/ios/JSI/EXJavaScriptRuntime.mm

```diff
@@ -19,35 +19,43 @@
 using namespace facebook;
 
 @implementation EXJavaScriptRuntime {
-  std::shared_ptr<jsi::Runtime> _runtime;
+  jsi::Runtime *_runtime;
   std::shared_ptr<react::CallInvoker> _jsCallInvoker;
+
+  /**
+   The unique pointer needs to be held when `EXJavaScriptRuntime` is an owner of the runtime (instead of the bridge).
+   This is necessary in tests and when we want to create independent runtime (when `init` without arguments is used).
+   */
+  std::unique_ptr<jsi::Runtime> _runtimeUniquePtr;
```

Comment:
i am not clear for the first time to see the comment and code flow.

maybe we could name it like `_runtimeUniquePtr -> _owningRuntimeForTesting` and `_runtime -> _bridgeRuntime`

other than that, it looks good to me.

Reply:
I'm against naming it "for testing" because it's only how we use it now, but in the future I can imagine we may want to spawn new runtimes like this (multithreading, worklets, workers, ...) in apps too.

The problem is that `make*Runtime` functions return a unique pointer, but we only need to get the raw pointer. We could do `_runtime = makeJSCRuntime().release()`, but then we're responsible for deleting the pointer, so we should also add this 👇 to prevent memory leaks.

```objc
- (void)dealloc
{
  delete _runtime;
}
```

On the other side, we can't do this when the `EXJavaScriptRuntime` is initialized through `initWithRuntime:callInvoker:` because the bridge may still hold a reference to this object.

To sum up, we need `_runtimeUniquePtr` only to keep the smart pointer and prevent it from deallocating the raw pointer.